File: src/main/org/nlogo/awt/Hierarchy.java
package org.nlogo.awt;

import java.awt.Component;
import java.awt.Container;
import java.awt.Frame;
import java.awt.Window;

public strictfp class Hierarchy {

  // this class is not instantiable
  private Hierarchy() { throw new IllegalStateException(); }

  /**
   * Returns the frame containing a component.
   */
  public static Frame getFrame(Component comp) {
    Component top = getTopAncestor(comp);
    if (top instanceof Frame) {
      return (Frame) top;
    } else {
      if (top instanceof Window &&
          top.getParent() != null) {
        return getFrame(top.getParent());
      }
      return null;
    }
  }

  /**
   * Returns the window containing a component.
   */
  public static Window getWindow(Component comp) {
    Component top = getTopAncestor(comp);
    if (top instanceof Window) {
      return (Window) top;
    } else {
      return null;
    }
  }

  public static Component getTopAncestor(Component comp) {
    Component top = comp;
    Container parent = top.getParent();
    while (!(top instanceof Window) && null != parent) {
      top = parent;
      parent = top.getParent();
    }
    return top;
  }

  public static boolean hasAncestorOfClass(Component component, Class<?> theClass) {
    return component != null &&
        (theClass.isInstance(component) ||
            hasAncestorOfClass(component.getParent(), theClass));
  }

  public static Container findAncestorOfClass(Component component, Class<?> theClass) {
    if (component == null) {
      return null;
    }
    if (theClass.isInstance(component)) {
      return (Container) component;
    }
    return findAncestorOfClass(component.getParent(), theClass);
  }

}

File: netlogo-gui/src/main/awt/Hierarchy.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.awt

import java.awt.{ Component, Frame, Window }

object Hierarchy {

  /**
   * Returns the frame containing a component.
   */
  def getFrame(comp: Component): Frame =
    getTopAncestor(comp) match {
      case top: Frame => top
      case top: Window if top.getParent != null =>
        getFrame(top.getParent)
      case _ => null
    }

  /**
   * Returns the window containing a component.
   */
  def getWindow(comp: Component): Window =
    getTopAncestor(comp) match {
      case top: Window => top
      case _ => null
    }

  def getTopAncestor(comp: Component): Component =
    Iterator.iterate(comp)(_.getParent)
      .takeWhile(_ != null)
      .collectFirst{case w: Window => w}
      .getOrElse(null)

  def hasAncestorOfClass(comp: Component, clazz: Class[_]) =
    findAncestorOfClass(comp, clazz).isDefined

  def findAncestorOfClass(comp: Component, clazz: Class[_]): Option[Component] =
    Iterator.iterate(comp)(_.getParent)
      .takeWhile(_ != null)
      .find(clazz.isInstance)

}
